tests/org.eclipse.swt.tests/JUnit Tests/org/eclipse/swt/tests/junit/Test_org_eclipse_swt_graphics_GC.java

```diff
@@ -273,14 +273,21 @@ public boolean isAutoScalable() {
     int canvasWidth = Math.round(bounds.width * targetScale);
     int canvasHeight = Math.round(bounds.height * targetScale);
     canvas.setSize(canvasWidth, canvasHeight);
+    final boolean[] paintListenerCalled = {false};
     canvas.addPaintListener(e -> {
         e.gc.drawImage(image, 0, 0, bounds.width, bounds.height,
                        0, 0, canvasWidth, canvasHeight);
+        paintListenerCalled[0] = true;
     });
 
     shell.open();
-    while (display.readAndDispatch()) {
-    }
+    long startTime = System.currentTimeMillis();
+    int timeoutMillis = 2000;
+	while (!paintListenerCalled[0] || System.currentTimeMillis() - startTime < timeoutMillis) {
+		while (display.readAndDispatch()) {
+			display.sleep();
+		}
```

Comment:
Without the condition inversion, this has the change to lock the application:
```suggestion
		if (!display.readAndDispatch()) {
			display.sleep();
		}
```